File: python/WolffIteration.py
import numpy as np

def WolffIteration(N, p, grid, adj):
    """
    Find a cluster according to the Wolff sampling rule - MATLAB-like implementation
    """
    import numpy as np
    
    # Random seed spin (0-indexed for Python)
    i = np.random.randint(0, N**2)
    
    # The cluster (initialized with seed)
    C = [i]
    
    # The frontier of spins
    F = [i]
    
    # Seed spin direction
    s = grid.flat[i]
    
    # Indicator function for cluster elements
    Ci = np.zeros(N**2, dtype=int)
    
    while len(F) > 0:
        # Get all neighbors of all frontier spins (vectorized)
        neighbors = adj[F].reshape(-1)  # Flattened array of all neighbors
        
        # Only choose neighbors parallel to the seed spin (vectorized)
        parallel_mask = [grid.flat[n] == s for n in neighbors]
        F = neighbors[parallel_mask]
        
        # Find elements that aren't in the cluster (using indicator arrays)
        Ci[C] = 1
        Fi = np.zeros(N**2, dtype=int)
        Fi[F] = 1
        
        # Elements in frontier but not in cluster
        F = np.where(Fi - Ci > 0)[0]
        
        # Apply probability filter (vectorized)
        r = np.random.random(len(F))
        F = F[r < p]
        
        # Add to cluster (vectorized)
        C.extend(F)
    
    return C, i

File: matlab/WolffIteration.m
function [C,i] = WolffIteration(N,p,grid,adj)
% Find a cluster, C, according the the Wolff sampling rule
%-------------------------------------------------------------------------------

i = randi(N^2); % random seed spin
C = i; % the cluster
F = i; % the frontier of spins
s = grid(i); % seed spin direction
Ci = zeros(N^2,1); % indicator function for cluster elements

while ~isempty(F)
    F = adj(F,:); % Compute the new neighboring spins
    F = F(grid(F(:)) == s); % only choose ones parallel to the seed spin
    % Find elements that aren't in the cluster
    Fi = zeros(N^2,1); % indicator function for the frontier spins
    Ci(C) = 1;
    Fi(F) = 1;
    F = find(Fi - Ci > 0);
    F = F(rand(1,length(F)) < p); % keep spins only with probability p
    C(end+1:end+length(F)) = F; % add to cluster
end

end
